components/tikv_util/src/quota_limiter.rs

```diff
@@ -0,0 +1,221 @@
+}
+
+impl<'a> Sample {
+    pub fn add_read_bytes(&mut self, bytes: usize) {
+        self.read_bytes += bytes;
+    }
+
+    pub fn add_write_bytes(&mut self, bytes: usize) {
+        self.write_bytes += bytes;
+    }
+
+    // Record the cpu time in the lifetime. Use this function inside code block.
+    // If `cputime_limiter` is not enabled, guard will do nothing when dropped.
+    pub fn observe_cpu(&'a mut self) -> CpuObserveGuard<'a> {
+        if self.enable_cpu_limit {
+            CpuObserveGuard {
+                timer: Some(ThreadTime::now()),
+                sample: self,
+            }
+        } else {
+            CpuObserveGuard {
+                timer: None,
+                sample: self,
+            }
+        }
+    }
+
+    fn add_cpu_time(&mut self, time: Duration) {
+        self.cpu_time += time;
+    }
+}
+
+pub struct CpuObserveGuard<'a> {
+    timer: Option<ThreadTime>,
+    sample: &'a mut Sample,
+}
+
+impl<'a> Drop for CpuObserveGuard<'a> {
+    fn drop(&mut self) {
+        if let Some(timer) = self.timer {
+            self.sample.add_cpu_time(timer.elapsed());
+        }
+    }
+}
+
+impl Default for QuotaLimiter {
+    fn default() -> Self {
+        Self {
+            cputime_limiter: Limiter::new(f64::INFINITY),
+            write_bandwidth_limiter: Limiter::new(f64::INFINITY),
+            read_bandwidth_limiter: Limiter::new(f64::INFINITY),
+        }
+    }
+}
+
+impl QuotaLimiter {
+    // 1000 millicpu equals to 1vCPU, 0 means unlimited
+    pub fn new(
+        cpu_quota: usize,
+        write_bandwidth: ReadableSize,
+        read_bandwidth: ReadableSize,
+    ) -> Self {
+        let cputime_limiter = if cpu_quota == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(cpu_quota as f64 * CPU_TIME_FACTOR * 1000_f64)
+        };
+
+        let write_bandwidth_limiter = if write_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(write_bandwidth.0 as f64)
+        };
+
+        let read_bandwidth_limiter = if read_bandwidth.0 == 0 {
+            Limiter::new(f64::INFINITY)
+        } else {
+            Limiter::new(read_bandwidth.0 as f64)
+        };
+
+        Self {
+            cputime_limiter,
+            write_bandwidth_limiter,
+            read_bandwidth_limiter,
+        }
+    }
+
+    // To generate a sampler.
+    pub fn new_sample(&self) -> Sample {
+        Sample {
+            read_bytes: 0,
+            write_bytes: 0,
+            cpu_time: Duration::ZERO,
+            enable_cpu_limit: !self.cputime_limiter.speed_limit().is_infinite(),
+        }
+    }
+
+    // To consume a sampler and return delayed duration.
+    // If the sampler is null, the speed limiter will just return ZERO.
+    pub async fn async_consume(&self, th: Sample) -> Duration {
+        let cpu_dur = self
+            .cputime_limiter
+            .consume_duration(th.cpu_time.as_micros() as usize);
```

Comment:
Is it a good idea to check `cpu_time` first?

Reply:
what's your point

Reply:
Do not call `consume_duration` if `cpu_time` is zero like read or write bytes.

Reply:
fixed